Tests/NIOTests/ChannelTests.swift

```diff
@@ -1716,4 +1716,143 @@ public class ChannelTests: XCTestCase {
             XCTFail("unexpected error: \(error)")
         }
     }
+
+    func testChannelReadsDoesNotHappenAfterRegistration() throws {
+        class ReadDoesNotHappen: ChannelInboundHandler {
+            typealias InboundIn = Any
+            private let hasRegisteredPromise: EventLoopPromise<()>
+            private let hasUnregisteredPromise: EventLoopPromise<()>
+            enum State {
+                case start
+                case registered
+                case active
+                case read
+            }
+            private var state: State = .start
+            init(hasRegisteredPromise: EventLoopPromise<()>, hasUnregisteredPromise: EventLoopPromise<()>) {
+                self.hasRegisteredPromise = hasRegisteredPromise
+                self.hasUnregisteredPromise = hasUnregisteredPromise
+            }
+            func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+                XCTAssertEqual(.active, self.state)
+                self.state = .read
+                #if !os(Linux)
+                XCTFail("read has happened and shouldn't have")
+                #endif
+            }
+            func channelActive(ctx: ChannelHandlerContext) {
+                XCTAssertEqual(.registered, self.state)
+                self.state = .active
+                #if !os(Linux)
+                XCTFail("shouldn't have been activated")
+                #endif
+            }
+            func channelRegistered(ctx: ChannelHandlerContext) {
+                XCTAssertEqual(.start, self.state)
+                self.state = .registered
+                self.hasRegisteredPromise.succeed(result: ())
+            }
+            func channelUnregistered(ctx: ChannelHandlerContext) {
+                self.hasUnregisteredPromise.succeed(result: ())
+            }
+        }
+
+        let group = MultiThreadedEventLoopGroup(numThreads: 2)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+        let serverEL = group.next()
+        let clientEL = group.next()
+        precondition(serverEL !== clientEL)
```

Comment:
@weissi just make this an `XCTAssertTrue` ?

Reply:
the problem is that if I do this, the test would just hang forever if something goes wrong (as `XCTAssert`s trigger _after_ the test has finished)

Reply:
I see... ok